lib/deimos/activerecord/batch_consumption.rb

```diff
@@ -0,0 +1,139 @@
+          zip(metadata[:keys]).
+          map { |p, k| [k, p] }
+
+        Deimos.instrument('ar_consumer.consume_batch', topic: metadata[:topic], messages: messages) do
+          slices = _slice_batch(messages)
+
+          # The entire batch should be treated as one transaction so that if
+          # any message fails, the whole thing is rolled back
+          ::ActiveRecord::Base.transaction do
+            slices.each do |slice|
+              removed, upserted = slice.partition { |_, v| v.nil? }
+
+              upsert_records(upserted) unless upserted.empty?
+              remove_records(removed) unless removed.empty?
+            end
+          end
+        end
+      end
+
+    protected
+
+      # Upsert any non-deleted records
+      # @param records [Array<Array>] List of [key, value] pairs for a group of
+      # non-tombstone records.
+      def upsert_records(records)
+        key_cols = key_columns(records)
+
+        messages = records.map { |k, v| record_attributes(k, v) }
+
+        # If record_attributes indicated no record, skip it
+        messages.compact!
+
+        options = if key_cols.empty?
+                    {} # Can't upsert with no key, just do regular insert
+                  else
+                    {
+                      on_duplicate_key_update: {
+                        # conflict_target must explicitly list the columns for
+                        # Postgres and SQLite
+                        conflict_target: key_cols,
+                        columns: :all
+                      }
+                    }
+                  end
+
+        @klass.import!(messages, options)
+      end
+
+      # Delete any records with a tombstone.
+      # @param records [Array<Array>] List of [key, nil] pairs for a group of
+      # deleted records.
+      def remove_records(records)
+        clause = deleted_query(records)
+
+        clause.delete_all
+      end
+
+      # Get unique key for the ActiveRecord instance from the incoming key.
+      # Override this method (with super) to customize the set of attributes that
+      # uniquely identifies each record in the database.
+      # @param key [Hash] The decoded key.
+      # @return [Hash] The key attributes.
+      def record_key(key)
+        decoded_key = decode_key(key)
+
+        if decoded_key.nil?
+          {}
+        elsif decoded_key.is_a?(Hash)
+          @key_converter.convert(decoded_key)
+        else
+          { @klass.primary_key => decoded_key }
+        end
+      end
+
+      # Create an ActiveRecord relation that matches all of the passed
+      # records. Used for bulk deletion.
+      # @param records [Array<Array>] List of [key, nil] pairs.
+      # @return ActiveRecord::Relation Matching relation.
+      def deleted_query(records)
+        keys = records.
+          map { |k, _| record_key(k) }.
+          reject(&:empty?)
+
+        keys.reduce(@klass.none) do |query, key|
+          query.or(@klass.unscoped.where(key))
```

Comment:
Can we not use `or`? This will crash on older versions of Rails. If it's a single field (and I think we can assume that case, allowing people to override this query where it isn't) you can just use `@klass.unscoped.where(key_field: decoded_keys)`. It would mean revamping how `record_key` works but I think that's probably better in any case - the keys of that query (the fields to check) should be identical for all records in the same topic.